### PR TITLE
[4.0] Colorpicker

### PR DESCRIPTION
**Pull request by brianteeman**
With the dropping of support for IE11 we can now finally start to use more browser native functionality. This PR completely drops the existing javascript solutions for a colour picker and replaces them all with a pure native html 5 colour picker. Unlike the current (broken) colourpickers this requires almost zero support or technical skills to maintain.

## Javascript (ES5, ES6, jQuery)
No javascript is used!!
 
## Accessibility
The colourpicker is completely accessible across all devices and browsers and when using any assistive technology. We don't need to do anything to maintain this accessibility as the browsers do it for us.

## Translation
The colourpicker is completely localised across all devices and browsers. We don't need to do anything to maintain the translations as the browsers do it for us.

## Colourpicker display
No code - it's all from the browser. This ensures a familiar ui. As more software drops IE support it will become even more familiar as everyone stops using external scripts

## Field Display
As well as adding some styling to the button to make it visibly consistent I have added a text representation of the color - see ToDo for further improvement suggestions

### Breaking change
Only supports RGB and HSL colours. RGBA is not supported **but** it has been broken for a long time with no sign of any interest in fixing it

## Advantages
 - almost zero maintenance
 - removes a jquery dependency
 - removes a 3pd script
 - removes (unmaintainable) code
 - familiar ui
 - completely accessible
 - responsive
 - great on mobile
 - zero translation
 
## Disadvantages
 - Does not support transparency or opacity
 - UI is not consistent across devices

## ToDo - (help would be awesome)
 [ ] The code blocks in the color field plugin need updating 
 [ ] Text representation of the selected colour only updates on save not change/selection
 [ ] It would be a bonus if the text representation of the colour was a text input so users could enter the value directly

### Testing instructions

1. Apply this PR
2. Composer install
2. npm install
3.  Add the following to any form xml (I used administrator\templates\atum\templateDetails.xml )
`
	<field
		name="color"
		type="color"
		label="Color test field"
		default="#000000"
		filter="color"
	/>
`
4. OR create a custom colour field

### CODE REVIEW

File: libraries/src/Form/Field/ColorField.php [hunk 1]
```diff
@@ -184,11 +84,12 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
 	 */
 	protected function getInput()
 	{
-		// Switch the layouts
-		$this->layout = $this->control === 'simple' ? $this->layout . '.simple' : $this->layout . '.advanced';
+	if (empty($this->layout))
```
Comment (Quy) [inline libraries/src/Form/Field/ColorField.php]:
```suggestion
		if (empty($this->layout))
```

File: libraries/src/Form/Field/ColorField.php [hunk 2]
```diff
@@ -204,106 +105,12 @@ protected function getLayoutData()
...
-			$color = 'none';
-		}
-		elseif ($color['0'] != '#' && $this->format == 'hex')
-		{
-			$color = '#' . $color;
-		}
-
-		// Assign data for simple/advanced mode
-		$controlModeData = $this->control === 'simple' ? $this->getSimpleModeLayoutData() : $this->getAdvancedModeLayoutData($lang);
+		$color = '#' . $color;
```
Comment (Quy) [inline libraries/src/Form/Field/ColorField.php:108]: If it is blank, still preappend `#`?

File: layouts/joomla/form/field/color.php [hunk 3]
```diff
@@ -49,18 +49,17 @@
  * @var   array    $colors          The specified colors
  */
 
-$class    = ' class="custom-select ' . trim($class) . '"';
+$class    = ' class="custom-select color' . trim($class) . '"';
```
Comment (dgrammatiko) [inline layouts/joomla/form/field/color.php:52]: Why are you removing the simple colour version? This is effectively a different field than the picker you want to change?

Comment (brianteeman) [reply layouts/joomla/form/field/color.php:52]: why maintain two pickers. especially as it was never documented that there were two

Comment (dgrammatiko) [reply layouts/joomla/form/field/color.php:52]: A. you can't just remove things without some deprecation period
B. Pretty sure there are people still using that

**Comment by ciar4n** [comment]
> RGBA is not supported but it has been broken for a long time with no sign of any interest in fixing it

How is it broken? Personally I use the format="rgba" attribute on the J3 picker quite extensively and have had no reported issues.

**Comment by brianteeman** [comment]
There are multiple open issues for it

**Comment by ciar4n** [comment]
> There are multiple open issues for it

I can only find one related issue and that is unconfirmed (https://github.com/joomla/joomla-cms/issues/19187).

**Comment by C-Lodder** [comment]
I not against this change, but bare in mind that on Android browsers (Chrome, Edge and Opera), it does work, however you cannot input your own HLS or RGB values, nor does it provide you the value. I don't have an iPhone, so can't test on there but assume it would be the same case.

Because of this, I'd strongly suggest having a separate text input next to the colour picker button showing the value of the selected colour, which would only require a few lines of JS

As for the opacity, there are already talks about it: https://github.com/w3c/html/issues/1422

A separate input range can be added in the same layout to support the opacity if need be.

Just something to think about.

**Comment by brianteeman** [comment]
> however you cannot input your own HLS or RGB values

You can but only from the sliders

> I'd strongly suggest having a separate text input next to the colour picker button showing the value of the selected colour, which would only require a few lines of JS

I agree and it would be useful on all devices thats why I put it on the TODO list

> ToDo - (help would be awesome)
[ ] The code blocks in the color field plugin need updating
[ ] Text representation of the selected colour only updates on save not change/selection
[ ] It would be a bonus if the text representation of the colour was a text input so users could enter the value directly

**Comment by dgrammatiko** [comment]
@brianteeman according to the n-2 I think you have to hold this pr for at least a year. Safari introduced color picker in this major release...

**Comment by viocassel** [comment]
I think the [browser native color picker](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/color) is not enough. Too limited functionality (for now) ...

How about replacing an existing [jQuery MiniColors](https://github.com/claviska/jquery-minicolors) with [Pickr](https://github.com/Simonwep/pickr)? Flat, simple, responsive and hackable Color-Picker library. No dependencies, no jQuery. Compatible with all CSS Frameworks e.g. Bootstrap, Materialize. Supports alpha channel, rgba, hsla, hsva and more!

image

**Comment by richard67** [comment]
@viocassel Pickr looks good and seems to be a currently maintained repo. But regarding people it seems to be a one man show. Can we be sure it will be maintained also in future?

**Comment by brianteeman** [comment]
Unfortunately pickr is not accessible at all and its not translatable

**Comment by Fedik** [comment]
Currently the color field support 2 mode: `simple` and `hue`. 
What if introduce another mode (eg `native`) and use it by default? basically it just a layout switching.
And who have used a `simple` mode will continue to use it without problem.

**Comment by ghost** [comment]
@viocassel please mail me at wohlkoenig.franz@gmail.com

**Comment by brianteeman** [comment]
closing due to lack of interest

**Comment by simonwep** [comment]
@brianteeman i18n support is [in progress](https://github.com/Simonwep/pickr/tree/i18n) (Different branch, based on kw-pairs) but currently not implemented in any release. I may not add completely new features in the future but I'll work on bugs and keep it up-to-date as good as possible. 

Stumbled across this PR while writing [this comment](https://github.com/joomla/joomla-cms/issues/27538#issuecomment-581957290) regarding #27538.

**Comment by ceford** [comment]
I noticed that the Color field type in Article Fields does not work. Not complaining - I don't want to use it - just noting the fact.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25162">issues.joomla.org/tracker/joomla-cms/25162</a>.</sub>

**Comment by roland-d** [comment]
@brianteeman  Any interest in fixing the conflicts and picking this up or should the PR be closed?

**Comment by brianteeman** [comment]
@roland-d no, no interest at all. seems joomla isnt really interested in removing jquery after all or in helping to get this finished

**Pull request closed without merging**